core/sync/index.js

```diff
@@ -64,13 +72,140 @@ const shouldAttemptRetry = (change /*: MetadataChange */) => {
+  } else if (!metadata.wasSynced(doc)) {
+    return { type: 'ADD', side: outdatedSide }
+  } else {
+    try {
+      const old = oldFromDoc(doc, outdatedSide.name)
+      if (old) {
+        if (
+          metadata.isFile(old) &&
+          metadata.isFile(doc) &&
+          metadata.sameFile(old, doc)
+        ) {
+          return { type: 'NULL' }
+        } else if (
+          metadata.isFolder(old) &&
+          metadata.isFolder(doc) &&
+          metadata.sameFolder(old, doc)
+        ) {
+          return { type: 'NULL' }
+        } else {
+          // We take as granted that doc and old have the same doc type (i.e.
+          // file or folder).
+          return { type: 'EDIT', side: outdatedSide }
+        }
+      } else {
+        return { type: 'ADD', side: outdatedSide }
+      }
+    } catch (err) {
+      return { type: 'EDIT', side: outdatedSide }
+    }
+  }
+}
+
+// Compare 2 Changes to determine which one should be propagated first.
+// It is the kind of function that could be passed to `Array.sort()`. It should
+// return:
+// - a negative value when the first Change argument should be propagated
+//   before the second Change argument
+// - a positive value when the second Change should be propagated before the
+//   first one
+// - 0 when the order of the Changes should be kept
+const compareChanges = (
+  { operation: opA, doc: docA } /*: Change */,
+  { operation: opB, doc: docB } /*: Change */
+) => {
+  // Operations on different sides don't have dependencies for now. This will
+  // probably be revisited in the future.
+  // SKIP and NULL operations don't have a side since we only update the PouchDB
+  // record and thus don't have dependencies.
+  if (opA.side != null && opB.side != null && opA.side === opB.side) {
+    if (opA.type === opB.type) {
+      // Handle same operation parent change before child changes
+      if (docB.path.startsWith(docA.path + sep)) return -1
+      else if (docA.path.startsWith(docB.path + sep)) return 1
+    }
+
+    if (opA.type === 'DEL' && opB.type === 'MOVE' && docB.moveFrom != null) {
+      // Handle parent deletion after child move outside of parent
+      if (
+        docB.moveFrom.path.startsWith(docA.path + sep) &&
+        !docB.path.startsWith(docA.path + sep)
+      ) {
+        return 1
+      }
+    }
+    if (opB.type === 'DEL' && opA.type === 'MOVE' && docA.moveFrom != null) {
+      // Handle parent deletion after child move outside of parent
+      if (
+        docA.moveFrom.path.startsWith(docB.path + sep) &&
+        !docA.path.startsWith(docB.path + sep)
+      )
+        return -1
+    }
+    if (opA.type === 'ADD' && opB.type === 'MOVE') {
+      // Handle child addtion after parent move
+      if (docA.path.startsWith(docB.path + sep)) return 1
+    }
+    if (opB.type === 'ADD' && opA.type === 'MOVE') {
+      // Handle child addtion after parent move
+      if (docB.path.startsWith(docA.path + sep)) return -1
+    }
```

Comment:
You don't have condition when both `opA` and `opB` are `'Add'` to ensure that the parent is created before the child? I think of this scenario:

1. the cozy-desktop is offline
2. we create the folder `A` on the cozy
3. we create the folder `B` on the cozy inside `A`
4. we rename the folder `A` to `C`
5. the desktop is now back online, and will see the creation of `C/B` before the creation of `C`

Reply:
And what about a file moved (opA) to a directory that is renamed (opB)?

Reply:
The first scenario (i.e. both operations are `ADD`) is covered by the same operation conditional block: https://github.com/cozy-labs/cozy-desktop/pull/2119/files#diff-7a081eeb18a2cc3146ed8c339c9b93edf2feafda1c0c53e1a7c03c53d89aaf7eR162

The second scenario will also enter the same operation conditional block and if the file is moved to the directory after it has been renamed then we should propagate the folder renaming first.
Otherwise the order will stay unchanged.